### DISCUSSION TITLE
Assessment of AI Tools in Development Workflow

### DISCUSSION THREAD

**Discussion by marjo-luc** [NASA-AMMOS/slim#91] [Ideas]
Integrating AI tools into the development workflow has been a topic that has gained a lot of attention recently. This may be a good opportunity to assess how/if these tools can be leveraged on projects like SLIM. 

Some areas that may benefit:

**Documentation**
Documentation tends to be a time-intensive activity and is often overlooked by projects. Consistent styling and terminology across a body of documentation is often of concern as well.
- Project Documentation
   Standard artifacts such as README's, FAQ's, or contributing guidelines are good candidates for AI-generated content.
- Code Documentation
   Exploring the efficacy of AI-generated code comments may be of interest as this may help developers better navigate 
   unfamiliar code bases and reduce onboarding time.

**Code Generation**
Developers are expected to work with a variety of ever-changing technology stacks. Working with technologies developers are not fluent in may affect their velocity. Leveraging AI tools for "smaller" implementation tasks may help free up time for developers to focus on "larger" implementation tasks. 

Code generation tasks that may benefit the most are:
- Configuration
   Dockerfiles, server configs, generic yamls, etc.
- Scripting
   Generating boiler plate code for classes or other objects as well as implementation of simple algorithms.
- Code Debugging
   Can these tools support developers in debugging existing codebases?

There are many AI-based tools on the market; two that have gained traction are [ChatGPT](https://openai.com/blog/chatgpt/) and [Copilot](https://github.com/features/copilot).

**Comment by perryzjc** [comment]
I feel that's an excellent idea that we could add to SLIM. I have been using the combination of ChatGPT and Copilot for a long time, which definitely makes my workflow much more efficient.

Whether it's writing documentation, generating code, debugging, or configuring, I've tried it all. Although sometimes these tools give the wrong answer, they often give a good starting point and save a lot of repetitive work.

**Comment by riverma** [comment]
Thank you so much for broaching this fascinating topic @marjo-luc and thanks for the comment @perryzjc! We live in exceptional times. At the minimum - this discussion itself could be useful for developers to be cognizant of - that tools like this even exist. 

Do you all feel that pointing / training developers in the SLIM community to potentially use these tools (and perhaps sharing some examples) is something that's needed? Or is there a gap we should fill in terms of _making_ a process improvement tool / template / automation to leverage these AI tools? Curious on your thoughts!

**Reply by marjo-luc** [reply]
I think we should assess how well these AI tools generate the content before we think about integration. Encouraging members of the community to explore them might be a good first step in better understanding what the limits of these technologies are and whether or not they are well-suited for our use cases.

I've been looking into [OpenAI's API](https://platform.openai.com/overview), in particular, and will share my findings.

**Reply by riverma** [reply]
Great point @marjo-luc - I'll keep this thread open to point folks to your idea. Let's evolve it organically, and yes please share your findings here as you dig deeper. Will do likewise.

**Comment by riverma** [comment]
@marjo-luc - take a look at wrapper tools that simplify the ChatGPT interface as well: https://github.com/TheR1D/shell_gpt
